### PR TITLE
Allow iterating over a set

### LINKED ISSUE #2023: Allow iterating over a set
### Describe the feature
It should be possible to iterate over a set.

### Test Code
Provide code which does not currently work but which should do when this issue is fixed:
```python
a = {1,3,4}
sum_a = 0
for ai in a:
     sum_a += ai
```

### Proposed Solution
This should be implemented using `IteratorType` and `Iterable`. The code should be flexible enough to make it easy to do the same for dicts (dict_keys/dict_values) later.

In C the code should be:
```c
c_for_each(Dummy_0000, set_int, a)
{
    ai = *Dummy_0000.ref;
    sum_a += ai;
}
```
In Fortran the code should be:
```fortran
      iter = a % begin()
      last = a % end()
      do while (iter /= last)
         ai = iter%of()
         sum_a = sum_a + ai
         call iter%next()
      end do
```

### PR DESCRIPTION
**Pull request by EmilyBourne**
Add support for iterating over a set. Fixes #2023

**Commit Summary**
- Add class type information for `zip`
- Complete definition of `IteratorType` following documentation
- Add printing support for iterating over a set
- Add iterator tests for sets

### CODE REVIEW

File: pyccel/ast/low_level_tools.py [hunk 1]
```diff
@@ -42,6 +42,39 @@
...
+    def order(self):
+        """
+        The data layout ordering in memory.
+
+        Indicates whether the data is stored in row-major ('C') or column-major
+        ('F') format. This is only relevant if rank > 1. When it is not relevant
+        this function returns None.
+        """
+        return None
+
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/low_level_tools.py:77]: /bot accept
Order is included for completeness

File: pyccel/codegen/printing/fcode.py [hunk 2]
```diff
@@ -2565,27 +2566,40 @@ def _print_FunctionalFor(self, expr):
...
-        if expr.iterable.num_loop_counters_required:
-            self.scope.insert_variable(index)
 
-        target   = index
-        my_range = expr.iterable.get_range()
+        if isinstance(iterable_type, (DictType, HomogeneousSetType)):
+            if isinstance(iterable.iterable, Variable):
+                suggested_name = iterable.iterable.name + '_'
+            else:
+                suggested_name = ''
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:2577]: /bot accept
This is expected. Ideally this fallback case should not be used

File: pyccel/ast/low_level_tools.py [hunk 3]
```diff
@@ -42,6 +42,39 @@ def iterable_type(self):
...
+    @property
+    def rank(self):
+        """
+        Number of dimensions of the object.
+
+        Number of dimensions of the object. If the object is a scalar then
+        this is equal to 0.
+        """
+        return 0
```
Comment (jalalium) [inline pyccel/ast/low_level_tools.py:65]: Is this always the case? What if we iterate over a list of lists?

Comment (EmilyBourne) [reply pyccel/ast/low_level_tools.py:65]: The iterator isn't the object inside the container. It is just a container containing a pointer to that object. It still needs unpacking. The only thing you can do with an STC iterator is `elem = iter->ref`. This means that you can never do `iter[i]` so I don't think it needs to be considered as a multi-dimensional object

File: tests/epyccel/test_epyccel_sets.py [hunk 4]
```diff
@@ -520,3 +520,33 @@ def union_int():
     pyccel_result = epyccel_func()
     python_result = union_int()
     assert python_result == pyccel_result
+
+def test_set_iter(language):
```
Comment (jalalium) [inline tests/epyccel/test_epyccel_sets.py:524]: This doesn't seem to work for  `C` and `Fortran`. There is a pytest param on line 488 that seems to restrict language to Python. I think handling iteration for all languages is within the scope of this PR.

**Review by jalalium — CHANGES_REQUESTED** [review]

File: tests/epyccel/test_epyccel_sets.py (hunk 4 above)
Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_sets.py:524]: The param on line 488 is a parametrize decorator. It only applies to the function that it is decorating

Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_sets.py:524]: Here is proof that this works for C and Fortran:
image

Comment (jalalium) [reply tests/epyccel/test_epyccel_sets.py:524]: I apologize profusely for the confusion.
I tested this and it did not work:
```python
if __name__ == '__main__':
    a = {1,2,3}
    ss = 0
    for x in a:
        ss += x
    print(ss)
```
that's why i thought that the tests did not run.
but then, why the following works but not the former:
```python
def f():
    a = {1,2,3}
    ss = 0
    for x in a:
        ss += x
    return ss

if __name__ == '__main__':
    print(f())
```

Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_sets.py:524]: What error are you seeing? This code seems to work fine for me

Comment (jalalium) [reply tests/epyccel/test_epyccel_sets.py:524]: So sorry, I had to delete __pyccel__, something was cached and was showing this error:
```
ERROR at code generation stage
pyccel:
 |error [codegen]: boo.py| Don't know how to index int type
Pyccel has encountered syntax that has not been implemented yet. Please create an issue at https://github.com/pyccel/pyccel/issues and provide a small example of your problem. Do not forget to specify your target language.
```

Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_sets.py:524]: That is the error on the devel branch. Very strange if you're seeing it here. I would guess that you maybe didn't have pyccel installed in editable mode but if that's the case then I've no idea how you've been developing ^^

Comment (jalalium) [reply tests/epyccel/test_epyccel_sets.py:524]: I do have it installed in editable mode. The issue had to do with the `__pyccel__` folder. It does not seem related but deleting it fixed the issue.

**Review by jalalium — APPROVED** [review]
Well done! So sorry for the oversight!

**Pull request merged**